### PR TITLE
[SofaKernel] ISSofa merge: Topology

### PR DESCRIPTION
**Pull request by guparan**
This PR will be further discussed during STC#3.

### New features
- UncoupledConstraintCorrection : add option to disable handling of topological changes for compliance vector
- Container : add accessor to the topology type
- SubsetMapping : add resizeToModel flag in SubsetMapping to enable resizing of the destination MechanicalState to the size of indices
- BaseContext : add new method getActiveMeshTopology to get relevant topology, taking `BaseMaping::sameTopology()` into account
- PointSetTopology : add some unit tests

### Improvements
- each time a topological change sequence is detected in BarycentricMapperTriangleSetTopology the mapping is initialized
- PointSetTopologyContainer maintains a list of point indices, accessible through the data points
- Topology : Factorize the last method specific to each topological elements in data containers

### Cleans
- MechanicalObject size and link to topology is now exposed as Data and Link, using the new getActiveMeshTopology() method. Set the link manually and/or set useTopology="false" to override

### Bugfixes
- apply data container updates before mapping updates
- use existing parameter handleTopologyChange in SubsetMapping
- slightly cleanup topology change handling method for triangle mappers
- check of upper topology based on the emptiness of the upper container
- crash in `MeshTopology::init()` for non manifold meshes
- re-enable Propagate of topological changes inbetween triangle creation and removal in `TriangleSetTopologyModifier::addRemoveTriangles(),` as it is required in cases involving chains of topology mapping and topology data containers depending on each other 

______________________________________________________

This PR: 
- [ ] builds with SUCCESS for all platforms on the CI.
- [x] does not generate new warnings nor unit test failures.
- [x] does not break existing scenes.
- [x] does not break API compatibility.
- [x] has been reviewed and agreed to be transitional.
- [x] is more than 1 week old.

**Reviewers will merge only if all these checks are true.**

### CODE REVIEW

**Comment by damienmarchal** [comment]
[build-ci]

**Comment by damienmarchal** [comment]
[ci-build]

File: SofaKernel/framework/sofa/core/objectmodel/BaseContext.cpp [hunk 1]
```diff
@@ -239,12 +239,26 @@ core::topology::Topology* BaseContext::getTopology() const
...
+core::topology::BaseMeshTopology* BaseContext::getLocalMeshTopology() const
+{
+    return this->get<sofa::core::topology::BaseMeshTopology>(Local);
+}
+
+/// Mesh Topology that is relevant for this context
+/// (within it or its parents until a mapping is reached that does not preserve topologies).
+core::topology::BaseMeshTopology* BaseContext::getActiveMeshTopology() const
+{
+    return this->get<sofa::core::topology::BaseMeshTopology>(Local);
```
Comment (epernod) [inline SofaKernel/framework/sofa/core/objectmodel/BaseContext.cpp:259]: Is it normal that `getLocalMeshTopology` and `getActiveMeshTopology` return exactly the same?

File: SofaKernel/modules/SofaBaseMechanics/BarycentricMapping.inl [hunk 2]
```diff
@@ -3727,6 +3727,82 @@ void BarycentricMapping<TIn, TOut>::handleTopologyChange ( core::topology::Topol
     reinit(); // we now recompute the entire mapping when there is a topologychange
 }
 
+#ifdef BARYCENTRIC_MAPPER_TOPOCHANGE_REINIT
```
Comment (epernod) [inline SofaKernel/modules/SofaBaseMechanics/BarycentricMapping.inl:3730]: The typdef `BARYCENTRIC_MAPPER_TOPOCHANGE_REINIT` is never defined in this PR. In which case should it be used?

File: SofaKernel/framework/sofa/core/objectmodel/BaseContext.cpp (hunk 1 above)
Comment (epernod) [reply SofaKernel/framework/sofa/core/objectmodel/BaseContext.cpp:259]: Should `GetActiveMeshTopology` replace current `GetMeshTopology` ?

**Comment by thomas-lemaire** [comment]
check that it works with DAGNode

**Comment by epernod** [comment]
All questions have been answered during STC. Ok to go as soon as possible as more modifications from Insimo are coming on the topology.

**Comment by guparan** [comment]
It seems that this PR is not building with SofaCUDA plugin.
See [ubuntu_clang-3.4_options](https://ci.inria.fr/sofa-ci/job/ubuntu_clang-3.4_options/4513/) and [ubuntu_clang-3.4_options-debug](https://ci.inria.fr/sofa-ci/job/ubuntu_clang-3.4_options-debug/193/) builds after merge.

**Pull request merged**